sdk/core/core/src/az_log.c

```diff
@@ -27,38 +27,61 @@ void az_log_set_callback(az_log_message_fn az_log_message_callback)
   _az_log_message_callback = az_log_message_callback;
 }
 
-void _az_log_write(az_log_classification classification, az_span message)
+// _az_log_write_engine is a function private to this .c file; it contains the code to handle
+// _az_log_should_write & _az_log_write.
+//
+// If log_it is false, then the function returns true or false indicating whether the message
+// should be logged (without actually logging it). if log_it is false, then the function returns
+// true or false indicating whether the message should be logged (without actually logging it).
+//
+// If log_it is true, then the function logs the message (if it should) and returns true or
+// false indicating whether it was logged.
+static bool _az_log_write_engine(bool log_it, az_log_classification classification, az_span message)
 {
-  if (_az_log_message_callback != NULL && _az_log_should_write(classification))
-  {
-    _az_log_message_callback(classification, message);
-  }
-}
+  // Copy the volatile fields to local variables so that they don't change within this function
+  az_log_message_fn const callback = _az_log_message_callback;
+  az_log_classification const* const classifications = _az_log_classifications;
```

Comment:
@JeffreyRichter , what if we read callback, then you change callback, then you change classifications, then we read classifications? Would we end up invoking the old callback with the potentially unexpected type of classification?
Do we need a mutex to read here?

Reply:
Yes, there is a race here but the race is expected since the functions to set the classification list and callback function are separate functions. So, a customer shouldn't have an expectation that they are treated as an atomic unit. We could have a single function to set both and then take a lock around the setting/getting but then we'd require the mutex and I'd love it if we can get it out of the PAL layer and so I'm trying to avoid more reasons to keep it.

Right now, we only need it for client secret credential and we may not need that.

Reply:
We can make it in a way that if you don't have a PAL layer, you probably don't have multiple threads as well, and in that case we compile without the mutex. It can be as simple as `#ifdef`, we also can optionally make these non-volatile if single threaded.